File: apps/py-cp-sat/ch04-puzzles/src/py_cp_sat_ch04/send_more_money.py
"""SEND + MORE = MONEY via CP-SAT.

Each letter is a distinct digit 0-9. S and M can't be zero (no leading zeros).
The puzzle has exactly one solution: 9567 + 1085 = 10652.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from ortools.sat.python import cp_model

_OK = (cp_model.OPTIMAL, cp_model.FEASIBLE)

_LETTERS = ("S", "E", "N", "D", "M", "O", "R", "Y")


@dataclass(frozen=True)
class SendMoreMoneyResult:
    """The unique letter-to-digit assignment."""

    status: str
    assignment: dict[str, int] = field(default_factory=dict)


def solve_send_more_money() -> SendMoreMoneyResult:
    """Solve the classic SEND + MORE = MONEY cryptarithm.

    Expected: S=9, E=5, N=6, D=7, M=1, O=0, R=8, Y=2.
    """
    model = cp_model.CpModel()
    variables = {letter: model.new_int_var(0, 9, letter) for letter in _LETTERS}

    model.add_all_different(list(variables.values()))
    model.add(variables["S"] >= 1)
    model.add(variables["M"] >= 1)

    send = (
        1000 * variables["S"]
        + 100 * variables["E"]
        + 10 * variables["N"]
        + variables["D"]
    )
    more = (
        1000 * variables["M"]
        + 100 * variables["O"]
        + 10 * variables["R"]
        + variables["E"]
    )
    money = (
        10000 * variables["M"]
        + 1000 * variables["O"]
        + 100 * variables["N"]
        + 10 * variables["E"]
        + variables["Y"]
    )
    model.add(send + more == money)

    solver = cp_model.CpSolver()
    solver.parameters.random_seed = 42
    status = solver.solve(model)
    if status not in _OK:
        return SendMoreMoneyResult(status=solver.status_name(status))
    return SendMoreMoneyResult(
        status=solver.status_name(status),
        assignment={k: int(solver.value(v)) for k, v in variables.items()},
    )

File: apps/kt-cp-sat/ch04-puzzles/src/main/kotlin/io/vanja/cpsat/ch04/SendMoreMoney.kt
package io.vanja.cpsat.ch04

import io.vanja.cpsat.*

/**
 * SEND + MORE = MONEY — the classic cryptarithmetic puzzle.
 *
 * Each distinct letter stands for a distinct digit (0..9). Leading digits
 * (S, M) are non-zero. The addition must hold.
 *
 * Unique solution:
 *   S=9 E=5 N=6 D=7  M=1 O=0 R=8 Y=2
 *   9567 + 1085 = 10652
 */

public data class SendMoreMoneySolution(
    val digits: Map<Char, Long>,
    val send: Long,
    val more: Long,
    val money: Long,
)

public fun solveSendMoreMoney(seed: Int = 42): SendMoreMoneySolution? {
    val letters = "SENDMORY".toList()
    val vars = mutableMapOf<Char, IntVar>()

    val model = cpModel {
        // Leading digits must be >= 1; others >= 0. Whole alphabet shares 0..9.
        for (c in letters) {
            val lo = if (c == 'S' || c == 'M') 1 else 0
            vars[c] = intVar(c.toString(), lo..9)
        }
        allDifferent(vars.values.toList())

        val s = vars['S']!!; val e = vars['E']!!; val n = vars['N']!!; val d = vars['D']!!
        val m = vars['M']!!; val o = vars['O']!!; val r = vars['R']!!; val y = vars['Y']!!

        val send = 1000 * s + 100 * e + 10 * n + d
        val more = 1000 * m + 100 * o + 10 * r + e
        val money = 10000 * m + 1000 * o + 100 * n + 10 * e + y

        constraint { +(send + more eq money) }
    }

    val res = model.solveBlocking { randomSeed = seed }
    val assignment = when (res) {
        is SolveResult.Optimal -> res.values
        is SolveResult.Feasible -> res.values
        else -> return null
    }
    val digits = letters.associateWith { assignment[vars[it]!!] }
    val sendV = 1000 * digits['S']!! + 100 * digits['E']!! + 10 * digits['N']!! + digits['D']!!
    val moreV = 1000 * digits['M']!! + 100 * digits['O']!! + 10 * digits['R']!! + digits['E']!!
    val moneyV = 10000 * digits['M']!! + 1000 * digits['O']!! + 100 * digits['N']!! +
        10 * digits['E']!! + digits['Y']!!
    return SendMoreMoneySolution(digits, sendV, moreV, moneyV)
}

public fun renderSendMoreMoney(sol: SendMoreMoneySolution): String = buildString {
    val d = sol.digits
    appendLine("  SEND  = ${sol.send}")
    appendLine("+ MORE  = ${sol.more}")
    appendLine("-------")
    appendLine("= MONEY = ${sol.money}")
    appendLine()
    appendLine("letters:")
    for (c in "SENDMORY") appendLine("  $c = ${d[c]}")
}
